### PR TITLE
Cargo extension for compiling Rust to MASM

### LINKED ISSUE #60: Cargo extension for compiling Rust code to MASM
In the first version of the extension, we will support only library targets because #59 is a blocker for app targets.

Commands:
* Create a new cargo project. Set `#no_std` required stuff in `lib.rs` (panic handler, etc.) and release profile options (size optimization, panic=abort, etc.)
* Compile the cargo project to MASM.

### PR DESCRIPTION
**Pull request by greenhat**
Close #60 

This PR introduces a cargo extension in the `cargo-miden` crate, which supports the following commands:
* `miden build` - compiles the selected target to MASM
* `miden new` - creates a new project from a template. The template is downloaded from a git repository.

Keep in mind that this cargo extension is planned to be the cargo-component
replacement when it comes to building/publishing/etc the Wasm/Miden components.
This should explain some implementation details. The overarching idea is
to introduce specific to Miden subcommands (e.g., `new`) and pass the
rest of the subcommands to cargo (e.g., `build`) and process the produced artifacts.

If the `wasm32-was` target is not present, it will be installed.

### CODE REVIEW

File: midenc-driver/src/commands/compile/stages/rewrite.rs
Comment (greenhat) [inline midenc-driver/src/commands/compile/stages/rewrite.rs:29]: Otherwise, `ApplyRewritesStage` crashes on `matches.get_flag` when arg matches are not provided in `Session`. The same applies to the second fix below. I made these fixes, assuming that `Session.matches` is not required. Let me know if I'm missing something.

File: cargo-ext/src/compile.rs
Comment (greenhat) [inline cargo-ext/src/compile.rs:85]: If I specify `project_type = ProjectType::Library`, the driver emits a program because `LinkerStage` produces a program if the output file type is MASM or MASL. See https://github.com/0xPolygonMiden/miden-ir/blob/29ba023b0f50b6c7afd205fe7c3e0ca6c81aa449/midenc-driver/src/commands/compile/stages/link.rs#L24

**Comment by greenhat** [comment]
@bobbinth Could you please create a new repo `miden-project-template` (or another name), for the template of the Miden project to be used by the cargo extension when creating a new project with `cargo miden new` command? I temporarily made https://github.com/greenhat/miden-project-template, but I cannot transfer it because github, for some reason, requires that I have a right to create a repo in the target organization. I'll copy the repo content ASAP. Thank you!

**Comment by bobbinth** [comment]
> @bobbinth Could you please create a new repo `miden-project-template` (or another name), for the template of the Miden project to be used by the cargo extension when creating a new project with `cargo miden new` command?

What kind of projects would this be for? For example, if this is to create smart contracts (i.e., account code), it should probably be called `miden-account-template`, and then in the future we'd have also `miden-note-template` (or something like this).

One other question: do we need to create a separate repo per command? Or could we create one repo (e.g., `miden-rust-templates`) and then put all templates into that repo?

**Comment by bitwalker** [comment]
> What kind of projects would this be for? For example, if this is to create smart contracts (i.e., account code), it should probably be called miden-account-template, and then in the future we'd have also miden-note-template (or something like this).

I think this is probably just for a simple Rust project template initially.

> One other question: do we need to create a separate repo per command? Or could we create one repo (e.g., miden-rust-templates) and then put all templates into that repo?

We need a repo per template I believe, so if we have three project templates, that's three separate repos. I'm not sure if we can put all templates in a single repo or not, @greenhat would be able to say for sure.

**Comment by greenhat** [comment]
> > What kind of projects would this be for? For example, if this is to create smart contracts (i.e., account code), it should probably be called miden-account-template, and then in the future we'd have also miden-note-template (or something like this).
> 
> I think this is probably just for a simple Rust project template initially.

Yes, so far, it's just compiling a `lib` target to the MASM module. No smart contract stuff yet. I assume we want to jump on account code right away and get back to dealing with generic Miden VM programs/libs later. So, I either evolve this template or create a new one for the account code.
 
> > One other question: do we need to create a separate repo per command? Or could we create one repo (e.g., miden-rust-templates) and then put all templates into that repo?
> 
> We need a repo per template I believe, so if we have three project templates, that's three separate repos. I'm not sure if we can put all templates in a single repo or not, @greenhat would be able to say for sure.

Good question! `cargo-generate` supports multiple templates in a single repo. And although initially, I wanted to have a single repo per template, I cannot find a good reason for it. It's not public facing, meaning no one will clone it, and the code cannot be built by cargo anyway. I think it makes sense to have a single repo for all templates. `miden-rust-templates` sounds good to me.

**Comment by bobbinth** [comment]
@greenhat - I've created [rust-templates](https://github.com/0xPolygonMiden/rust-templates) and you should have write access to it. Let me know if anything else is needed.

**Comment by greenhat** [comment]
@bitwalker ready for review

File: cargo-ext/src/new_project.rs
Comment (bitwalker) [inline cargo-ext/src/new_project.rs:28]: I don't think you need a library for this, `path.canonicalize()` (or alternatively `std::fs::canonicalize`) produce a canonical, absolute path.

File: midenc-session/src/lib.rs
Comment (bitwalker) [inline midenc-session/src/lib.rs:288]: The output file path is guaranteed to always be a file path, not a directory, so this check is unnecessary.

**Review by bitwalker — CHANGES_REQUESTED** [review]
Nice work!

There's a few things I'd like to change about the way the extension works, and a few cosmetic things, but overall this looks good to me. I think we can probably have this merged tomorrow, but let me know if some of the changes I've suggested will complicate that and we can discuss - but I can either pick up where you leave off tomorrow, or we can just wait an extra day or whatever.

File: cargo-ext/src/compile.rs
Comment (bitwalker) [reply cargo-ext/src/compile.rs:31]: NOTE: My later comment that goes into how we determine what target to build and all that using Cargo metadata covers this as well, but my point about following the conventions of `cargo build` still holds in terms of choosing defaults IMO.

Comment (greenhat) [reply cargo-ext/src/compile.rs:31]: > If we want to allow specifying a target, we should follow the conventions of `cargo build` I think, to avoid having different defaults than expected.

I like the idea of mirroring `cargo build` options and semantics! It got me thinking about taking the approach of `cargo component` for building the Wasm component - just pass all the options to `cargo` and pick up Wasm files afterward. See  https://github.com/bytecodealliance/cargo-component/blob/87949555d067e92a1872c03daa272794a6c0f6a5/src/lib.rs#L132-L180
I think we can take the same approach for our `build`(`compile`) command. This means for our `build` command, we'll have only `--miden-target` option for VM vs. rollup (I'm not even sure we need that), and the rest will be passed to `cargo build` as is adding `--target wasm32-wasi` if it's not present. I believe the Wasm component binary alone would suffice to figure out if Miden program or library is being built.

File: cargo-ext/src/cli_commands.rs
Comment (greenhat) [reply cargo-ext/src/cli_commands.rs:32]: > I don't think this flag is needed, probably - we can assume `--bin` is the default build type (just like `cargo build`), and use the default binary target for the crate being compiled (using the crate name as the output file name).
> 
> We probably want to add a `--lib` flag though, to build a library target rather than a binary target. We could also flip the defaults so that `--lib` is the default and `--bin` must be specified, but either way I don't expect that multiple binary targets is likely to be a common case.
> 
> What's your feeling on it?

See https://github.com/0xPolygonMiden/compiler/pull/61#discussion_r1415552505

File: cargo-ext/src/compile.rs
Comment (greenhat) [reply cargo-ext/src/compile.rs:19]: Sounds good! We'll need options to specify dependencies, etc.. Still, I think `components` and `default-target` options will add extra complexity for the newcomers and might do more harm than good. Given that they are not strictly necessary for the build process, we should get back to them later when we have a clear picture of the smart contract development workflow (crate structure, testing, etc.). 
To drive the build process, I'd instead take the `cargo-component` approach. See my comment - https://github.com/0xPolygonMiden/compiler/pull/61#discussion_r1415552505

File: cargo-ext/src/cli_commands.rs
Comment (greenhat) [reply cargo-ext/src/cli_commands.rs:26]: Nope, I copied it from `midenc`. Fixed. I do like using `name` though. It allows renaming the enum variant without breaking the cmd line options.

File: cargo-ext/src/cli_commands.rs
Comment (greenhat) [reply cargo-ext/src/cli_commands.rs:27]: Agree. Renamed.

File: cargo-ext/src/cli_commands.rs
Comment (greenhat) [reply cargo-ext/src/cli_commands.rs:36]: I agree. Will do.

File: cargo-ext/src/compile.rs
Comment (bitwalker) [reply cargo-ext/src/compile.rs:10]: I generally agree with you, and I'm not really a fan of the approach used by the other Miden projects, so I've compromised by using something that is more of a blend: if it's in the same namespace, combine it, e.g. `use a::b::Foo;` and `use a::b::Bar` would get combined, but `use a::b::Foo` and `use a::c::Baz` would be separated. That generally avoids most conflicts due to changes in imports, while collapsing them considerably.

If this was my own project, I'd probably switch to grouping by item, but I wanted to at least make the attempt to follow the Miden guidelines to some degree here.

If there's a config file we can add to the project root that sets that rust-analyzer config option locally for the project, feel free to add it. Definitely annoying to have to change it globally just for the Miden projects, so if there's any way to avoid that, I'm open to it.

File: cargo-ext/src/new_project.rs
Comment (greenhat) [reply cargo-ext/src/new_project.rs:28]: Good catch! Thanks! Fixed.

File: cargo-ext/src/compile.rs
Comment (bitwalker) [reply cargo-ext/src/compile.rs:19]: > Still, I think components and default-target options will add extra complexity for the newcomers and might do more harm than good

To be clear, those are not needed with a default project setup, only for more complex workspaces which contain a blend of Miden and non-Miden crates. For simple cases, we can just assume all crates are Miden crates, and use the default targets implied by the Cargo configuration. It is very unlikely that most crates will have anything other than a single lib or bin target anyway.

File: cargo-ext/src/compile.rs
Comment (bitwalker) [reply cargo-ext/src/compile.rs]: I think I'm fine with following `cargo component` here, though I think it remains to be seen whether our more complex build process forces us to be more restrictive than they are. That said, it feels like a better starting point for sure, and will probably get us quite far before we need to do anything special.

File: midenc-compile/src/stages/parse.rs
Comment (greenhat) [reply midenc-compile/src/stages/parse.rs:41]: Good idea! Will do.

File: Cargo.toml
Comment (greenhat) [reply Cargo.toml]: Done.

File: Cargo.toml
Comment (greenhat) [reply Cargo.toml]: Good catch! I probably initially thought of using it in the integration tests. Removed.

File: midenc-session/src/lib.rs
Comment (bitwalker) [reply midenc-session/src/lib.rs:288]: Hmm, the OutputFile API is designed to always return a file path, so if it's returning a directory there's a somewhere. I'll take a closer look

File: cargo-ext/src/compile.rs
Comment (greenhat) [reply cargo-ext/src/compile.rs]: Got it. It seems this import granularity is `Module`. I made an issue - https://github.com/0xPolygonMiden/compiler/issues/78

**Comment by greenhat** [comment]
@bitwalker I finished implementing all the review notes and it's ready for review. Please check the updated PR description for a summary.

**Review by bitwalker — APPROVED** [review]
LGTM. I'll need to look into the `OutputFile` thing I mentioned, but I'm not going to hold this PR up for that

**Pull request merged**
